### ISSUE TITLE
`std_wildcard_imports` to lint against `use std::mod::*`

### ISSUE THREAD

**Issue by tgross35** [rust-lang/rust-clippy#13961]
### What it does

This is a version of [`wildcard_imports`](https://rust-lang.github.io/rust-clippy/master/index.html#wildcard_imports) that affects only imports from the standard crates `std`, `core`, `proc_macro`, `alloc`, and `test`.

### Advantage

`wildcard_imports` is in pedantic, which seems reasonable for most crates. However, using glob imports from the standard library means that upgrading Rust version can break the build. https://github.com/rust-lang/rust/issues/134707 and https://github.com/rust-lang/rust/issues/134714 are recent example.

I am proposing `std_wildcard_imports` as a separate lint from `wildcard_imports` so it can get enabled by default, likely in `clippy::style`.

### Drawbacks

Confusion between `wildcard_imports` and `std_wildcard_imports`, noise on existing code, sometimes `use proc_macro::*` or similar is fine for testing.

### Example

Lint on glob imports:

```rust
use std::mem::*; // error
use core::sync::*; // error
use proc_macro::*; // error
```

However, using glob imports from prelude modules should not raise this warning:

```rust
use std::io::prelude::*; // ok
use std::prelude::rust_2021:* // unlikely, but still ok
```

Non-`std` crates don't raise this lint

```rust
use regex::*; // ok
```

**Comment by tgross35** [comment]
Quick usage search https://github.com/search?q=lang%3Arust+%2Fuse+%28std%7Ccore%7Cproc_macro%29.*%3A%3A%5C*%2F&type=code. Most of those are `io::prelude`, but there are at least a handful of results on the first few pages that would be linted.

**Comment by jieyouxu** [comment]
There's a caveat that the prelude modules like `std::io::prelude::*` can still *participate* in breakage (but not from std adding items), in that a user may glob import another external crate in addition to one of these std preludes, then if that *external crate* adds something that collides with the (practically frozen) std preludes, you can still end up with glob-vs-glob ambiguous. Less likely of course, agreed with not linting on std preludes.

**Comment by tgross35** [comment]
We have hit a few more cases where this lint would have been very useful. Since there is an existing lint that can be referenced, it seems like a reasonable first issue.

@rustbot label +E-help-wanted +good-first-issue
